File: blf_lib/src/blf/chunks/halo3/v06481_06_11_17_1330_alpha_release/s_blf_chunk_packed_game_variant.rs
use std::io::{Read, Seek, Write};
use binrw::{BinRead, BinResult, BinWrite, BinWriterExt, Endian};
use serde::{Deserialize, Serialize};
use blf_lib::blam::halo3::v06481_06_11_17_1330_alpha_release::game::game_engine_variant::c_game_variant;
use blf_lib::io::bitstream::{c_bitstream_reader, c_bitstream_writer, e_bitstream_byte_order};
use blf_lib_derivable::blf::chunks::BlfChunkHooks;
use blf_lib_derive::BlfChunk;

#[derive(BlfChunk,Default,PartialEq,Debug,Clone,Serialize,Deserialize)]
#[Header("gvar", 2.1)]
pub struct s_blf_chunk_packed_game_variant
{
    pub game_variant: c_game_variant,
}

impl BlfChunkHooks for s_blf_chunk_packed_game_variant {}

impl BinRead for s_blf_chunk_packed_game_variant {
    type Args<'a> = ();

    fn read_options<R: Read + Seek>(reader: &mut R, endian: Endian, args: Self::Args<'_>) -> BinResult<Self> {
        let mut buffer = Vec::<u8>::new();
        reader.read_to_end(&mut buffer)?;

        let mut bitstream = c_bitstream_reader::new_with_legacy_settings(buffer.as_slice(), e_bitstream_byte_order::_bitstream_byte_order_big_endian);
        bitstream.begin_reading();

        let mut packed_game_variant = Self::default();

        packed_game_variant.game_variant.decode(&mut bitstream)?;

        Ok(packed_game_variant)
    }
}

impl BinWrite for s_blf_chunk_packed_game_variant {
    type Args<'a> = ();

    fn write_options<W: Write + Seek>(&self, writer: &mut W, endian: Endian, args: Self::Args<'_>) -> BinResult<()> {
        let mut bitstream = c_bitstream_writer::new_with_legacy_settings(0x264, e_bitstream_byte_order::_bitstream_byte_order_big_endian);
        bitstream.begin_writing();

        self.game_variant.encode(&mut bitstream)?;

        bitstream.finish_writing();
        writer.write_ne(&bitstream.get_data()?)
    }
}

File: blf-ts/src/chunks/haloreach/v12065_11_08_24_1738_tu1actual/s_blf_chunk_packed_game_variant.ts
import {
  c_bitstream_reader,
  c_bitstream_writer,
  e_bitstream_byte_order,
} from "../../../bitstream";
import { blf, BLFChunkBase } from "../../../blf_chunk";
import type { c } from "@craftycodie/cstruct";
import { BlfError } from "../../../error";
import { c_game_variant } from "../../../blam/haloreach/v12065_11_08_24_1738_tu1actual/game/c_game_variant";

/**
 * Reach matchmaking game variant chunk (`gvar` 54.1).
 *
 * Payload is the gametype bitstream only (no mpvr hash/header wrapper).
 */
@blf.chunk("gvar", 54.1)
export class s_blf_chunk_packed_game_variant extends BLFChunkBase {
  game_variant = new c_game_variant();

  static create(
    game_variant: c_game_variant,
  ): s_blf_chunk_packed_game_variant {
    const chunk = new s_blf_chunk_packed_game_variant();
    chunk.game_variant = game_variant;
    return chunk;
  }

  read_body(payload: Uint8Array, _endian: c.Endian): void {
    if (payload.length === 0) {
      throw new BlfError("gvar chunk payload is empty");
    }

    const bitstream = c_bitstream_reader.new(
      payload,
      e_bitstream_byte_order._bitstream_byte_order_big_endian,
    );
    bitstream.begin_reading();
    this.game_variant = new c_game_variant();
    this.game_variant.decode(bitstream);
    bitstream.finish_reading();
  }

  write_body(_endian: c.Endian): Uint8Array {
    const bitstream = c_bitstream_writer.new(
      0x5028,
      e_bitstream_byte_order._bitstream_byte_order_big_endian,
    );
    bitstream.begin_writing();
    this.game_variant.encode(bitstream);
    bitstream.finish_writing();
    return bitstream.get_data();
  }
}
